cylc/flow/network/graphql.py

```diff
@@ -128,200 +149,88 @@ def strip_null(data):
-                    continue
-                frag_def = self.fragment_defs[field.name.value]
-                frag_type = type_from_ast(self.schema, frag_def.type_condition)
-                if self.args_selection_search(
-                        frag_def.selection_set, variables,
-                        frag_type, arg_name, arg_value):
-                    return True
-                self.visited_fragments.add(frag_def.name)
-                continue
-            field_def = get_field_def(
-                self.schema, parent_type, field.name.value)
-            if field_def is None:
-                continue
-            arg_vals = get_argument_values(
-                field_def.args, field.arguments, variables)
-            if arg_vals.get(arg_name) == arg_value:
-                return True
-            if field.selection_set is None:
-                continue
-            if self.args_selection_search(
-                    field.selection_set, variables,
-                    get_named_type(field_def.type), arg_name, arg_value):
-                return True
-        return False
-
-
-def execute_and_validate_and_strip(
-    schema: 'GraphQLSchema',
-    document_ast: 'Document',
-    *args: Any,
-    **kwargs: Any
-) -> Union['ExecutionResult', Observable]:
-    """Wrapper around graphql ``execute_and_validate()`` that adds
-    null stripping."""
-    result = execute_and_validate(schema, document_ast, *args, **kwargs)
-    # Search request document to determine if 'stripNull: true' is set
-    # as and argument. It can not be done in the middleware, as they
-    # can be Promises/futures (so may not been resolved at this point).
-    variable_values = kwargs['variable_values'] or {}
-    doc_args = AstDocArguments(schema, document_ast, variable_values)
-    if doc_args.has_arg_val(STRIP_ARG, True):
-        if kwargs.get('return_promise', False) and hasattr(result, 'then'):
-            return result.then(null_stripper)  # type: ignore[union-attr]
-        return null_stripper(result)
-    return result
-
-
-class CylcGraphQLBackend(GraphQLBackend):
-    """Return a GraphQL document using the default
-    graphql executor with optional null-stripping of result.
-
-    The null value stripping of result is triggered by the presence
-    of argument & value "stripNull: true" in any field.
+    if is_awaitable(exe_result):
+        return async_next(strip_null, exe_result)
+    return strip_null(exe_result)
 
-    This is a modification of GraphQLCoreBackend found within:
-        https://github.com/graphql-python/graphql-core-legacy
-    (graphql-core==2.3.2)
 
-    Args:
-
-        executor (object): Executor used in evaluating the resolvers.
+class CylcVisitor(Visitor):
+    """Traverse graphql document/query to find an argument in a given state.
 
+    Find whether an argument is set to a specific value anywhere in the
+    document (i.e. 'strip_null' set to 'True'), and stop on the first
+    occurrence.
```

Comment:
I think either a link to https://github.com/graphql-python/graphql-core/blob/0107e309cde181035cc806a5abd0358c37924251/src/graphql/language/visitor.py#L171 or 

```python
    __doc__ = (
        """Your doc
        """
        + Visitor.__doc__
```

Might be useful.

Reply:
Are we to do that with every inherited class?.. Seems a little much

Well.. I would go for the latter, because then we don't need to version pin the link..

Reply:
```
    __doc__ = (
        """Search graphql document/query to find matching argument/value pair.
        Find whether an argument is set to a specific value anywhere in the
        document (i.e. 'strip_null' set to 'True'), and stop on the first
        occurrence.
        """
        + Visitor.__doc__
    )
```
gives
```
$ mypy
cylc/flow/network/graphql.py:161: error: Unsupported operand types for + ("str" and "None")  [operator]
cylc/flow/network/graphql.py:161: note: Right operand is of type "Optional[str]"
Found 1 error in 1 file (checked 209 source files)
```
any ideas?

Reply:
(happy to leave the dev to look up the docs here)

Reply:
Sorry - pointless convo.